### PR TITLE
Decouple driver from Cypress, use Assertable

### PR DESCRIPTION
**Pull request by ShellyDCMS**
Test Refactoring
1. All assertions are in the test rather than in the driver
2. separate give, when and get
3. use then and Assertable
4. use CypressHelper for decoupling from Cypress

### CODE REVIEW

File: projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts [hunk 1]
```diff
@@ -11,27 +11,28 @@ import {
   Output,
   SimpleChanges,
   ViewChild,
-} from '@angular/core';
-import { LngLatLike, Offset, Popup, PopupOptions } from 'maplibre-gl';
-import { MapService } from '../map/map.service';
-import { MarkerComponent } from '../marker/marker.component';
+} from "@angular/core";
```
Comment (HarelM) [inline projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts:14]: Can this changes be reverted? I think the quotes are OK, aren't they?

File: projects/showcase/cypress/e2e/language-switch.cy.ts [hunk 2]
```diff
@@ -1,49 +1,55 @@
...
       beforeEach(() => {
-        driver
-          .visitMapPage('/demo/language-switch')
-          .waitForMapToIdle()
-
-          // Start in French for the baseline
-          .when.clickLanguageButton('French')
-          .waitForMapToIdle()
-          .takeImageSnapshot();
+        ({ when, get, then } = new E2eDriver());
```
Comment (HarelM) [inline projects/showcase/cypress/e2e/language-switch.cy.ts:13]: I think there's a redundant `)` and `(` here, right?

Comment (HarelM) [reply projects/showcase/cypress/e2e/language-switch.cy.ts:13]: Also this is done in the context as well, so I'm not sure it's needed here...?

File: projects/showcase/cypress/e2e/popup.cy.ts [hunk 3]
```diff
@@ -1,26 +1,31 @@
...
+      when.visitMapPage("/demo/popup");
+      when.waitForMapToIdle();
     });
 
-    context('When the map loads', () => {
+    context("When the map loads", () => {
       it('Then I should see a popup dialog containing "Hello World !"', () => {
-        driver.assert.helloWorldPopupExists();
+        then(get.mglPopup()).shouldExist();
+        then(get.mapLibrePopUpContent()).should("have.text", "Hello world !");
```
Comment (HarelM) [inline projects/showcase/cypress/e2e/popup.cy.ts:18]: Doesn't the driver has a `shouldHaveText` method?

File: projects/showcase/cypress/support/e2e-driver.ts [hunk 4]
```diff
@@ -1,198 +1,179 @@
...
-    clickDisableTerrainControlButton: (): E2eDriver => {
-      cy.get('.maplibregl-ctrl-terrain-enabled').click();
-      return this;
-    },
-    clickHideControlsButton: (): E2eDriver => {
-      cy.get('button').contains('Hide Controls').click();
-      return this;
-    },
+  given = {
+    alertStub: () => cy.on("window:alert", this.helper.given.stub("alert")),
```
Comment (HarelM) [inline projects/showcase/cypress/support/e2e-driver.ts:62]: Should this `cy.on` be part of an extended helper?

File: projects/showcase/cypress/support/e2e-driver.ts [hunk 5]
```diff
@@ -1,198 +1,179 @@
...
-      return this;
-    },
-    helloWorldPopupExists: (): E2eDriver => {
-      cy.get('.maplibregl-popup').should('exist');
-      cy.get('.custom-popup-class1').should('exist');
-      cy.get('.custom-popup-class2').should('exist');
-      cy.get('.maplibregl-popup-content').within(() => {
-        cy.get('div').should('have.text', 'Hello world !');
+    resetConsoleWarnings: () => {
+      cy.get("@consoleWarnSpy").then((spy: any) => {
```
Comment (HarelM) [inline projects/showcase/cypress/support/e2e-driver.ts:68]: According to your architecture, there should not be calls to `cy` in the driver right?

File: projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts (hunk 1 above)
Comment (ShellyDCMS) [reply projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts:14]: It's done automatically by prettier....
Is there a prettier config file in this repo?

File: projects/showcase/cypress/e2e/language-switch.cy.ts (hunk 2 above)
Comment (ShellyDCMS) [reply projects/showcase/cypress/e2e/language-switch.cy.ts:13]: This re-assigning of destructed attributes, will not compile without ()
I don't see where it's done in the context, can you give me a permalink to line?

File: projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts (hunk 1 above)
Comment (ShellyDCMS) [reply projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts:14]: Maybe I'll just search and replace all " with ' in this repo?

File: projects/showcase/cypress/e2e/popup.cy.ts (hunk 3 above)
Comment (ShellyDCMS) [reply projects/showcase/cypress/e2e/popup.cy.ts:18]: You mean Assertable, right?
No, not yest, I'll add it.

File: projects/showcase/cypress/support/e2e-driver.ts (hunk 4 above)
Comment (ShellyDCMS) [reply projects/showcase/cypress/support/e2e-driver.ts:62]: Hmm..
Yes, we could add an extend helper to maplibre as well.

File: projects/showcase/cypress/support/e2e-driver.ts (hunk 5 above)
Comment (ShellyDCMS) [reply projects/showcase/cypress/support/e2e-driver.ts:68]: Yes, not quite sure what to do with this one, as well as the two commands that are in the commands file
https://github.com/maplibre/ngx-maplibre-gl/blob/3ea724c8ea03999a3ad83c3d9fba778faa5f64b9/projects/showcase/cypress/support/commands.ts#L45#L53

File: projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts (hunk 1 above)
Comment (ShellyDCMS) [reply projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts:14]: No, that's a bad idea...

Comment (ShellyDCMS) [reply projects/ngx-maplibre-gl/src/lib/popup/popup.component.ts:14]: Fixed it with lint

File: projects/showcase/cypress/support/e2e-driver.ts (hunk 5 above)
Comment (ShellyDCMS) [reply projects/showcase/cypress/support/e2e-driver.ts:68]: I moved it to MaplibreCypreeHelper

File: projects/showcase/cypress/e2e/language-switch.cy.ts (hunk 2 above)
Comment (HarelM) [reply projects/showcase/cypress/e2e/language-switch.cy.ts:13]: A few lines above this line.

File: projects/showcase/cypress/e2e/popup.cy.ts [hunk 6]
```diff
@@ -1,26 +1,31 @@
-import { E2eDriver } from '../support/e2e-driver';
+import { E2eDriver } from "../support/e2e-driver";
```
Comment (HarelM) [inline projects/showcase/cypress/e2e/popup.cy.ts:1]: @ShellyDCMS please run the linter again, seems like not all files are using the same rules on quotes, right?

File: projects/showcase/cypress/e2e/terrain-control.cy.ts [hunk 7]
```diff
@@ -1,49 +1,50 @@
...
+
+      beforeAndAfter();
 
       beforeEach(() => {
-        driver
-          .visitMapPage('/demo/terrain-control')
-          .waitForMapToIdle()
-          .assert.mapTerrainButtonIsOff()
-          .takeImageSnapshot();
+        ({ when, get, then } = new E2eDriver());
```
Comment (HarelM) [inline projects/showcase/cypress/e2e/terrain-control.cy.ts:13]: This line and line 7 seems identical, what am I missing?
This repeats in other tests as well.

File: projects/showcase/cypress/support/e2e-driver.ts [hunk 8]
```diff
@@ -1,198 +1,112 @@
...
-  private referenceImageBuffer: Buffer;
-  private alertStub: Cypress.Agent<Sinon.SinonStub<any[], any>>;
+  private helper = new MaplibreCypressHelper();
 
-  visitMapPage = (url: string): E2eDriver => {
-    cy.visit(url);
-    this.alertStub = cy.stub().as('alertStub');
-    return this;
+  beforeAndAfter = () => {
+    beforeEach(() => {});
```
Comment (HarelM) [inline projects/showcase/cypress/support/e2e-driver.ts:8]: What's the point in those if they are empty?

File: projects/showcase/cypress/support/e2e-driver.ts [hunk 9]
```diff
@@ -1,198 +1,112 @@
...
-  takeImageSnapshot = (): E2eDriver => {
-    this.initReferenceImage();
-    return this;
-  };
+    resetConsoleWarnings: () => this.helper.when.resetConsoleWarnings(),
+    waitForStreetsSpriteResponse: (url: string, method: string = "GET") =>
+      this.helper.when.waitForResponse("streets"),
+    waitForMapToIdle: () =>
+      this.helper.when.waitUntil(() =>
+        this.helper.get.bySelector("true", "data-idle")
```
Comment (HarelM) [inline projects/showcase/cypress/support/e2e-driver.ts:25]: this method looks odd `bySelector("true", "data-idle")` I would expect it to be: bySelector("data-idle", "true")

File: projects/showcase/cypress/support/e2e-driver.ts [hunk 10]
```diff
@@ -1,198 +1,112 @@
...
-      cy.get('.maplibregl-ctrl-fullscreen').should('not.exist');
-      return this;
-    },
-    customPopupContainsHello: (): E2eDriver => {
-      cy.get('@alertStub').should('have.been.calledOnceWith', 'Hello');
-      return this;
-    },
-  };
+  get = {
+    canvas: (): Cypress.Chainable<HTMLCanvasElement> =>
```
Comment (HarelM) [inline projects/showcase/cypress/support/e2e-driver.ts:69]: Is this used "outside" the driver? Seem like this should be internal...?

File: projects/showcase/cypress/support/e2e-driver.ts [hunk 11]
```diff
@@ -1,198 +1,112 @@
...
-  private toImageBitmapBuffer(canvas: HTMLCanvasElement) {
-    const base64 = canvas.toDataURL('image/png').replace(/data:.*;base64,/, '');
-    const buff = Cypress.Buffer.from(base64, 'base64');
-    const png = PNG.sync.read(buff as any);
-    return png.data;
-  }
+      return snapshot;
+    },
+
+    imageBitmapBuffer: (canvas: HTMLCanvasElement) => {
```
Comment (HarelM) [inline projects/showcase/cypress/support/e2e-driver.ts:87]: Same here, this looks like something that should not be exposed to the test code.

File: projects/showcase/cypress/support/maplibre-helper.ts [hunk 12]
```diff
@@ -0,0 +1,78 @@
+import { CypressHelper } from "@shellygo/cypress-test-utils";
+import { Assertable, then } from "@shellygo/cypress-test-utils/assertable";
+import pixelmatch from "pixelmatch";
+export class MapLibreAssertable<T> extends Assertable<T> {
```
Comment (HarelM) [inline projects/showcase/cypress/support/maplibre-helper.ts:4]: Shouldn't the assertable have its own file?

**Comment by HarelM** [comment]
I can't tell which comments were addressed and which weren't. I'm leaving it to you to find out.
Overall this looks great but it's hard to review due to all the extra noise...
Thanks!

File: projects/showcase/cypress/e2e/terrain-control.cy.ts (hunk 7 above)
Comment (ShellyDCMS) [reply projects/showcase/cypress/e2e/terrain-control.cy.ts:13]: This is recreating the Driver before each test. It is to make sure there are no side effects left between the execution of tests.

File: projects/showcase/cypress/support/e2e-driver.ts (hunk 8 above)
Comment (ShellyDCMS) [reply projects/showcase/cypress/support/e2e-driver.ts:8]: placeholders for others who may wish to add something in the future.  I think we should move the code from cypress commands into here.

File: projects/showcase/cypress/support/e2e-driver.ts (hunk 9 above)
Comment (ShellyDCMS) [reply projects/showcase/cypress/support/e2e-driver.ts:25]: The second parameter has a default (data-cy).
Maybe I should add another method doing the same thing with a clearer name.

File: projects/showcase/cypress/support/e2e-driver.ts (hunk 10 above)
Comment (ShellyDCMS) [reply projects/showcase/cypress/support/e2e-driver.ts:69]: Why internal? What if I want to assert things about the canvas in a test? That it exists? That it's visible?

File: projects/showcase/cypress/support/e2e-driver.ts (hunk 11 above)
Comment (ShellyDCMS) [reply projects/showcase/cypress/support/e2e-driver.ts:87]: Why?

File: projects/showcase/cypress/support/maplibre-helper.ts (hunk 12 above)
Comment (ShellyDCMS) [reply projects/showcase/cypress/support/maplibre-helper.ts:4]: Not sure. I was thinking all Maplibre's extensions to test utils can be here, so we don't have to go back and forth between source files. WDYT?

Comment (HarelM) [reply projects/showcase/cypress/support/maplibre-helper.ts:4]: If it was methods I would agree, but I prefer that every class would have its own file.

File: projects/showcase/cypress/support/e2e-driver.ts (hunk 11 above)
Comment (HarelM) [reply projects/showcase/cypress/support/e2e-driver.ts:87]: Doesn't sound like something that would make sense to "PM" if they were writing the tests. This should be the "language" of the driver, right?

File: projects/showcase/cypress/support/e2e-driver.ts (hunk 10 above)
Comment (HarelM) [reply projects/showcase/cypress/support/e2e-driver.ts:69]: Do you do it right now? If not, let's make it private. We can always make it public in the future.

File: projects/showcase/cypress/e2e/terrain-control.cy.ts (hunk 7 above)
Comment (HarelM) [reply projects/showcase/cypress/e2e/terrain-control.cy.ts:13]: This sounds like a smell. The beforeEach and afterEach should take care of it.

**Pull request merged**